# Ejercicio 1

In [8]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score

import warnings

import funciones_auxiliares_machine_learning as aux

warnings.filterwarnings('ignore')
plt.rcParams['figure.figsize'] = (12, 6)
plt.style.use('seaborn')

df = pd.read_csv('ames_housing.csv').drop(columns='Unnamed: 0')

# Ejercicio 2

In [3]:
import pickle

modelo_propio = pickle.load(open('javier-rojas-actividad07.sav', 'rb'))
X_train = pickle.load(open('javier-rojas-X-train-actividad07.sav', 'rb'))
y_train = pickle.load(open('javier-rojas-y-train-actividad07.sav', 'rb'))
X_test = pickle.load(open('javier-rojas-X-test-actividad07.sav', 'rb'))
y_test = pickle.load(open('javier-rojas-y-test-actividad07.sav', 'rb'))

In [5]:
aux.reportar_metricas_regresion(modelo_propio, X_test, y_test)

|     |      valores |
|:----|-------------:|
| MAE | 24419.6      |
| R2  |     0.757795 |


## Reportar metricas de modelo de companero TODO

# Ejercicio 3

In [6]:
random_forest = RandomForestRegressor(random_state=1)
random_forest.fit(X_train, y_train)

RandomForestRegressor(random_state=1)

In [7]:
aux.reportar_metricas_regresion(random_forest, X_test, y_test)

|     |      valores |
|:----|-------------:|
| MAE | 19723.4      |
| R2  |     0.859388 |


- La tabla comparativa de resultados son las siguientes:


Modelo|MAE|R2
------|---|--
Grilla|24420|0.76|
RandomForest|19723|0.85|

- Por lo tanto, se puede observar un drastico aumento en el desempeno del modelo (disminucion del error, y aumento del poder explicativo r2).

# Ejercicio 4

In [19]:
grilla = {
    'max_features': [None, 'log2', 'sqrt'],
    'n_estimators': range(20, 1001, 50)
}

random_forest_grilla = GridSearchCV(RandomForestRegressor(oob_score=True, bootstrap=True), param_grid=grilla,
                                    verbose=5, n_jobs=-1)
random_forest_grilla.fit(X_train, y_train)

Fitting 5 folds for each of 60 candidates, totalling 300 fits


GridSearchCV(estimator=RandomForestRegressor(oob_score=True), n_jobs=-1,
             param_grid={'max_features': [None, 'log2', 'sqrt'],
                         'n_estimators': range(20, 1001, 50)},
             verbose=5)

In [31]:
print(f'Los mejores parametros son: {random_forest_grilla.best_params_}')
print(f'El mejor desempeno es: {random_forest_grilla.best_score_}')
print(f'El mejor desempeno OOB es: {random_forest_grilla.best_estimator_.oob_score_}')

Los mejores parametros son: {'max_features': 'sqrt', 'n_estimators': 370}
El mejor desempeno es: 0.8517314409920766
El mejor desempeno OOB es: 0.8547396987887688


In [32]:
oob_score = random_forest_grilla.best_estimator_.oob_score_

## Graficar comportamiento del modelo TODO

# Ejercicio 5

In [36]:
random_optimo = RandomForestRegressor(n_estimators=370, max_features='sqrt', oob_score=True, bootstrap=True,
                                      random_state=1).fit(X_train, y_train)
aux.calcular_metricas_regresion(random_optimo, X_test, y_test)

valores
MAE  19449.119708
R2       0.857203

- La tabla resumen de todos los modelos son los siguientes:

Modelo|MAE|R2
------|---|--
Grilla|24420|0.76|
RandomForest|19723|0.859|
RandomForest Optimo|19449|0.857|

- Teniendo en cuenta estos resultados, el modelo optimo presenta el mae mas bajo y un r2 comparable al random forest sin optimizacion de hiperparametros. Por lo tanto, el modelo
optimizado presenta el mejor desempeno.